src/Horarium.InMemory/InMemoryRepository.cs

```diff
@@ -0,0 +1,156 @@
+
+            lock (jobWrapper.SyncRoot)
+            {
+                jobWrapper.Job.Status = JobStatus.Failed;
+                jobWrapper.Job.Error = error.Message + error.StackTrace;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task RemoveJob(string jobId)
+        {
+            _jobsStorage.TryRemove(jobId, out var _);
+
+            return Task.CompletedTask;
+        }
+
+        public Task RepeatJob(string jobId, DateTime startAt, Exception error)
+        {
+            if (!_jobsStorage.TryGetValue(jobId, out var jobWrapper))
+                return Task.CompletedTask;
+
+            lock (jobWrapper.SyncRoot)
+            {
+                jobWrapper.Job.Status = JobStatus.RepeatJob;
+                jobWrapper.Job.StartAt = startAt;
+                jobWrapper.Job.Error = error.Message + error.StackTrace;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task AddRecurrentJob(JobDb job)
+        {
+            bool Filter(JobDb x) =>
+                x.JobKey == job.JobKey &&
+                (x.Status == JobStatus.Executing || x.Status == JobStatus.Ready);
+
+            void Update(JobDb x)
+            {
+                x.Cron = job.Cron;
+                x.StartAt = job.StartAt;
+            }
+
+            var result = UpdateOneAtomic(Filter, Update);
+            if (result != null) return Task.CompletedTask;
+
+            var copy = JobInMemoryWrapper.CopyJob(job);
+            Update(copy);
+            
+            return AddJob(copy);
+        }
+
+        public Task AddRecurrentJobSettings(RecurrentJobSettings settings)
+        {
+            _recurrentJobSettingsStorage.AddOrUpdate(settings.JobKey, settings, 
+                (_, __) => settings);
+            
+            return Task.CompletedTask;
+        }
+
+        public Task<string> GetCronForRecurrentJob(string jobKey)
+        {
+            if (!_recurrentJobSettingsStorage.TryGetValue(jobKey, out var settings))
+                throw new Exception("Не найдены настройки для рекуррентного джоба");
+
+            return Task.FromResult(settings.Cron);
+        }
+
+        public Task<Dictionary<JobStatus, int>> GetJobStatistic()
+        {
+            var dict = _jobsStorage.Values.Select(x => x.Job)
+                .GroupBy(x => x.Status)
+                .ToDictionary(x => x.Key, g => g.Sum(x => 1));
+            
+            foreach (JobStatus jobStatus in Enum.GetValues(typeof(JobStatus)))
+            {
+                if (!dict.ContainsKey(jobStatus))
+                {
+                    dict.Add(jobStatus, 0);
+                }
+            }
+
+            return Task.FromResult(dict);
+        }
+
+        private JobDb UpdateOneAtomic(Func<JobDb, bool> filter, Action<JobDb> update)
+        {
+            foreach (var kvp in _jobsStorage)
+            {
+                var jobWrapper = kvp.Value;
+
+                if (filter(jobWrapper.Job))
```

Comment:
What about performance in this search method? Example 100 000 jobs.

Reply:
Yea, performance kinda sucks here, 'cause it's basically collscan.

There are 2 reasons why i implemented it this exact way:
1. How real is the case of 100k jobs? - i mean inMemory doesn't seem to be a good choice for such a huge workload anyway.
2. I might try to implement some index-like stuff, but then code will become waaay more complex, I'm afraid in order of magnitude more complex.

Sounds like a fun thing to try tho, will go for it

Reply:
Well if we go for performance-first (or at least more performant) implementation, i'll just reimplement everything and create another PR - i dont see how things can be done with this synchronization model - background thread processing concurrent FIFO queue of operations (wrapped by TaskCompletionSource) is what i'm thinking of. Does that sound ok to you? - if so, let's just close this one and wait for another try coming shortly

Reply:
I want to create load testing, after that we will see